lib/src/modules/apply.cc

```diff
@@ -0,0 +1,234 @@
+#include <pxp-agent/util/bolt_helpers.hpp>
+#include <pxp-agent/configuration.hpp>
+#include <boost/filesystem/path.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/tokenizer.hpp>
+#include <leatherman/json_container/json_container.hpp>
+#include <leatherman/file_util/file.hpp>
+
+namespace fs = boost::filesystem;
+namespace lth_file = leatherman::file_util;
+namespace lth_jc = leatherman::json_container;
+
+namespace PXPAgent {
+namespace Modules {
+
+    static const std::string apply_ACTION { "apply" };
+    static const std::string apply_ACTION_INPUT_SCHEMA { R"(
+    {
+        "type": "object",
+        "properties": {
+            "environment": {
+                "type": "string"
+            },
+            "version": {
+                "type": "string"
+            },
+            "catalog": {
+                "type": "object"
+            }
+        },
+        "required": ["environment", "version", "catalog"]
+    }
+    )" };
+
+    // TODO: Actually manage this file in packaging (or perhaps fetch it from puppetserver).
+    // For now just copy it to the cache dir each time unless there is already a copy
+    static const std::string RUBY_SHIM {
+R"(#! /opt/puppetlabs/puppet/bin/ruby
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'json'
+require 'puppet'
+require 'puppet/configurer'
+require 'puppet/module_tool/tar'
+require 'securerandom'
+require 'tempfile'
+
+## TODO: Option to read from a file is for debugging. Only read from stdin in production.
+args = JSON.parse(ARGV[0] ? File.read(ARGV[0]) : STDIN.read)
+
+# Create temporary directories for all core Puppet settings so we don't clobber
+# existing state or read from puppet.conf. Also create a temporary modulepath.
+# Additionally include rundir, which gets its own initialization.
+puppet_root = Dir.mktmpdir
+moduledir = File.join(puppet_root, 'modules')
+Dir.mkdir(moduledir)
+cli = (Puppet::Settings::REQUIRED_APP_SETTINGS + [:rundir]).flat_map do |setting|
+  ["--#{setting}", File.join(puppet_root, setting.to_s.chomp('dir'))]
+end
+cli << '--modulepath' << moduledir
+Puppet.initialize_settings(cli)
+
+# Avoid extraneous output
+Puppet[:report] = false
+
+# Make sure to apply the catalog
+Puppet[:noop] = args['_noop'] || false
+# TODO: enable apply setting
+#args['apply_settings'].each do |setting, value|
+#  Puppet[setting.to_sym] = value
+#end
+
+Puppet[:default_file_terminus] = :file_server
+
+exit_code = 0
+begin
+  # This happens implicitly when running the Configurer, but we make it explicit here. It creates the
+  # directories we configured earlier.
+  Puppet.settings.use(:main)
+
+  ## TODO: ENABLE PLUGINS
+  # Tempfile.open('plugins.tar.gz') do |plugins|
+  #   File.binwrite(plugins, Base64.decode64(args['plugins']))
+  #   user = Etc.getpwuid.nil? ? Etc.getlogin : Etc.getpwuid.name
+  #   Puppet::ModuleTool::Tar.instance.unpack(plugins, moduledir, user)
+  # end
+
+  env = Puppet.lookup(:environments).get('production')
```

Comment:
TODO: use env from shim arguments.

Reply:
I actually dont think this does anything. We set up an isolated environment for this execution but puppet needs a "real" environment, so we use `production`. I asked the bolt team yesterday in slack but it did not get any feedback.

Reply:
This name might end up in the report? We may have to make it not-production. Unfortunately Puppet has some weirdness about creating environments not named production out of thin air...

Reply:
I seem to have been able to get  `Puppet::Node::Environment.remote(args['environment'])` to work for everything. I will investigate that further.

Reply:
I dont see it in the report. I am really tempted to use the "battle tested" (at least in bolt) method of using `'production'` here.

Reply:
Yeah, I think it's fine to leave this with a comment about the discussion we've had here. Just something to describe why this says production and that we don't see anything in the report to make us think it needs to be different.